backend/app/notify/email.go

```diff
@@ -0,0 +1,380 @@
+}
+
+// verifyTmplData store data for verification message template execution
+type verifyTmplData struct {
+	User  string
+	Email string
+	Token string
+	Site  string
+}
+
+const defaultEmailTimeout = 10 * time.Second
+const defaultFlushDuration = time.Second * 30
+const defaultEmailTemplate = `{{.From}}{{if .To}} → {{.To}}{{end}}
+
+{{.Orig}}
+
+↦ <a href="{{.Link}}">{{if .PostTitle}}{{.PostTitle}}{{else}}original comment{{end}}</a>
+`
+
+const defaultEmailVerificationTemplate = `Confirmation for {{.User}} {{.Email}}, site {{.Site}}
+
+Token: {{.Token}}
+`
+
+//NewEmail makes new Email object, returns it even in case of problems
+// (e.MsgTemplate parsing error or error while testing smtp connection by credentials provided in params)
+func NewEmail(params EmailParams) (*Email, error) {
+	var err error
+	res := Email{EmailParams: params}
+	if res.TimeOut <= 0 {
+		res.TimeOut = defaultEmailTimeout
+	}
+	if res.FlushDuration <= 0 {
+		res.FlushDuration = defaultFlushDuration
+	}
+	if res.BufferSize <= 0 {
+		res.BufferSize = 1
+	}
+	if res.MsgTemplate == "" {
+		res.MsgTemplate = defaultEmailTemplate
+	}
+	if res.VerificationTemplate == "" {
+		res.VerificationTemplate = defaultEmailVerificationTemplate
+	}
+	// unbuffered send channel for sending messages to autoFlush goroutine
+	res.submit = make(chan emailMessage)
+
+	log.Printf("[DEBUG] create new email notifier for server %s with user %s, timeout=%s",
+		res.Host, res.Username, res.TimeOut)
+
+	// initialise templates
+	res.msgTmpl, err = template.New("messageFromRequest").Parse(res.MsgTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse message template")
+	}
+	res.verifyTmpl, err = template.New("messageFromRequest").Parse(res.VerificationTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse verification template")
+	}
+
+	// establish test connection
+	testSMTPClient, err := res.client()
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't establish test connection")
+	}
+	if err = testSMTPClient.Quit(); err != nil {
+		log.Printf("[WARN] failed to send quit command to %s:%d, %v", res.Host, res.Port, err)
+	}
+	if err = testSMTPClient.Close(); err != nil {
+		log.Printf("[WARN] can't close smtp connection, %v", err)
+	}
+	return &res, err
+}
+
+// Send email from request to request.parentUserEmail if it's set, otherwise do nothing and return nil, thread safe
+// do not returns sending error, except following cases:
+// 1. (likely impossible) template execution error from email message creation from request
+// 2. message dropped without sending in case of closed ctx
+func (e *Email) Send(ctx context.Context, req request) error {
+	if req.parentUserEmail == "" || req.parent.User == req.comment.User {
+		// don't send anything if there is no email to send information to
+		// or if user replied to his own comment
+		return nil
+	}
+	// start auto flush once, as this is the first moment we see the context from caller
+	e.once.Do(func() {
```

Comment:
just a general point: running goroutine from once could make testing harder. If there is higher level (service mentioned above) start of the gorotine is better to delegate to that level.

Reply:
Point taken. I've tried to do that in the first place but had to abandon this idea because we need to pass context, as stated in the comment. I see it making code flow more complex, but I found no better way to do it. Resolving.

Reply:
I'm not sure how this expected to work from ctx point of view. It passes ctx once to the goroutine and ignores for any other call